modules/puzzle/src/main/PuzzleApi.scala

```diff
@@ -112,4 +113,42 @@ private[puzzle] final class PuzzleApi(
         }
       }
   }
+
+  object learning {
+
+    def find(user: User): Fu[Option[Learning]] =
+      learningColl.find($doc(
+        Learning.BSONFields.id -> user.id
+      )).uno[Learning]
+
+    def add(l: Learning) = learningColl insert l void
+
+    def update(user: User, puzzle: Puzzle, data: DataForm.AttemptData) = 
+      if (data.isWin) solved(user, puzzle.id) else failed(user, puzzle.id)
+
+    def solved(user: User, puzzleId: PuzzleId) = learning find user flatMap {
+      case None => fuccess(none)
+      case Some(l) =>
+        learningColl.update(
+          $id(l.id),
+          $doc("$set" -> $doc(Learning.BSONFields.stack -> l.stack.filter(_ != puzzleId))))
+    }
+
+    def failed(user: User, puzzleId: PuzzleId) = learning find user flatMap {
+      case None => learning add Learning(user.id, List(puzzleId))
+      case Some(l) =>
+        learningColl.update(
+          $id(l.id),
+          $doc("$set" -> $doc(Learning.BSONFields.stack -> l.addPuzzle(puzzleId)))) 
```

Comment:
Perhaps, but if the id already exists in the collection it must be removed, which is what addPuzzle does.